Compilation fails on NetBSD (conflict with system-defined RANDOM_MAX)

**Godot version:**
3.2.2 and 3.2.3

**OS/device including version:**
NetBSD 9 AMD64

**Issue description:**
When I issue the compilation command scons platform=x11 or if I want to compile with llvm: scons platform=x11 use_llvm=yes, they both fail very early.

**Steps to reproduce:**
NetBSD has the proper versions for dependencies and I believe that I have all dependencies satisfied from the package collection for NetBSD, but trying to compile it fails. If I use gcc I get a mix of several very vague errors like one mentioning an expected ';' in one file. If I used llvm I get an expected unqualified-id error. It doesn't matter if use the git, the stable releases for 3.2.2 or 3.2.3, or the port from openbsd. They all fail.


Edit: 
I will admit I was using virtualbox to render NetBSD, but it's been perfect performance so I really don't want to disregard this as a VM issue rather than possible godot issue. 

edit 2: Here are the logs in text form:
```
gcc-
scons: Reading SConscript files ...
Enabling ALSA
Enabling PulseAudio
Checking for C header file mntent.h... (cached) no
scons: done reading SConscript files.
scons: Building targets ...

[Initial build] [94mCompiling [95m==> [93mplatform/x11/godot_x11.cpp[0m
In file included from platform/x11/godot_x11.cpp:33:0:
./core/math/random_pcg.h:70:24: error: expected unqualified-id before numeric constant
  static const uint64_t RANDOM_MAX = 0xFFFFFFFF;
                        ^
./core/math/math_funcs.h:50:24: error: expected unqualified-id before numeric constant
  static const uint64_t RANDOM_MAX = 0xFFFFFFFF;
                        ^
./core/math/math_funcs.h: In static member function 'static double Math::randd()':
./core/math/math_funcs.h:283:80: error: expected unqualified-id before numeric constant
  static _ALWAYS_INLINE_ double randd() { return (double)rand() / (double)Math::RANDOM_MAX; }
                                                                                ^
./core/math/math_funcs.h:283:80: error: expected ';' before numeric constant
In file included from ./core/math/vector2.h:34:0,
                 from ./core/math/rect2.h:34,
                 from ./core/math/transform_2d.h:34,
                 from ./core/os/input_event.h:34,
                 from ./core/os/main_loop.h:34,
                 from ./core/engine.h:35,
                 from ./core/os/os.h:34,
                 from platform/x11/context_gl_x11.h:38,
                 from platform/x11/os_x11.h:34,
                 from platform/x11/godot_x11.cpp:37:
./core/math/math_funcs.h:283:90: warning: statement has no effect [-Wunused-value]
  static _ALWAYS_INLINE_ double randd() { return (double)rand() / (double)Math::RANDOM_MAX; }
                                                                                          ^
In file included from platform/x11/godot_x11.cpp:33:0:
./core/math/math_funcs.h: In static member function 'static float Math::randf()':
./core/math/math_funcs.h:284:77: error: expected unqualified-id before numeric constant
  static _ALWAYS_INLINE_ float randf() { return (float)rand() / (float)Math::RANDOM_MAX; }
                                                                             ^
./core/math/math_funcs.h:284:77: error: expected ';' before numeric constant
In file included from ./core/math/vector2.h:34:0,
                 from ./core/math/rect2.h:34,
                 from ./core/math/transform_2d.h:34,
                 from ./core/os/input_event.h:34,
                 from ./core/os/main_loop.h:34,
                 from ./core/engine.h:35,
                 from ./core/os/os.h:34,
                 from platform/x11/context_gl_x11.h:38,
                 from platform/x11/os_x11.h:34,
                 from platform/x11/godot_x11.cpp:37:
./core/math/math_funcs.h:284:87: warning: statement has no effect [-Wunused-value]
  static _ALWAYS_INLINE_ float randf() { return (float)rand() / (float)Math::RANDOM_MAX; }
                                                                                       ^
scons: *** [platform/x11/godot_x11.x11.tools.64.o] Error 1
scons: building terminated because of errors.

LLVM-
scons: Reading SConscript files ...
Enabling ALSA
Enabling PulseAudio
Checking for C header file mntent.h... (cached) no
scons: done reading SConscript files.
scons: Building targets ...

[Initial build] [94mCompiling [95m==> [93mplatform/x11/godot_x11.cpp[0m
In file included from platform/x11/godot_x11.cpp:37:
In file included from platform/x11/os_x11.h:34:
In file included from platform/x11/context_gl_x11.h:38:
In file included from ./core/os/os.h:34:
In file included from ./core/engine.h:35:
In file included from ./core/os/main_loop.h:34:
In file included from ./core/os/input_event.h:34:
In file included from ./core/math/transform_2d.h:34:
In file included from ./core/math/rect2.h:34:
In file included from ./core/math/vector2.h:34:
In file included from ./core/math/math_funcs.h:35:
[1m./core/math/random_pcg.h:70:24: [0m[0;1;31merror: [0m[1mexpected member name or ';' after
      declaration specifiers[0m
        static const uint64_t RANDOM_MAX = 0xFFFFFFFF;
[0;1;32m        ~~~~~~~~~~~~~~~~~~~~~ ^
[0m[1m/usr/include/stdlib.h:188:20: [0m[0;1;30mnote: [0mexpanded from macro 'RANDOM_MAX'[0m
#define RANDOM_MAX      0x7fffffff      /* (((long)1 << 31) - 1) */
[0;1;32m                        ^
[0mIn file included from platform/x11/godot_x11.cpp:37:
In file included from platform/x11/os_x11.h:34:
In file included from platform/x11/context_gl_x11.h:38:
In file included from ./core/os/os.h:34:
In file included from ./core/engine.h:35:
In file included from ./core/os/main_loop.h:34:
In file included from ./core/os/input_event.h:34:
In file included from ./core/math/transform_2d.h:34:
In file included from ./core/math/rect2.h:34:
In file included from ./core/math/vector2.h:34:
[1m./core/math/math_funcs.h:50:24: [0m[0;1;31merror: [0m[1mexpected member name or ';' after
      declaration specifiers[0m
        static const uint64_t RANDOM_MAX = 0xFFFFFFFF;
[0;1;32m        ~~~~~~~~~~~~~~~~~~~~~ ^
[0m[1m/usr/include/stdlib.h:188:20: [0m[0;1;30mnote: [0mexpanded from macro 'RANDOM_MAX'[0m
#define RANDOM_MAX      0x7fffffff      /* (((long)1 << 31) - 1) */
[0;1;32m                        ^
[0mIn file included from platform/x11/godot_x11.cpp:37:
In file included from platform/x11/os_x11.h:34:
In file included from platform/x11/context_gl_x11.h:38:
In file included from ./core/os/os.h:34:
In file included from ./core/engine.h:35:
In file included from ./core/os/main_loop.h:34:
In file included from ./core/os/input_event.h:34:
In file included from ./core/math/transform_2d.h:34:
In file included from ./core/math/rect2.h:34:
In file included from ./core/math/vector2.h:34:
[1m./core/math/math_funcs.h:283:80: [0m[0;1;31merror: [0m[1mexpected unqualified-id[0m
  ...double randd() { return (double)rand() / (double)Math::RANDOM_MAX; }
[0;1;32m                                                            ^
[0m[1m/usr/include/stdlib.h:188:20: [0m[0;1;30mnote: [0mexpanded from macro 'RANDOM_MAX'[0m
#define RANDOM_MAX      0x7fffffff      /* (((long)1 << 31) - 1) */
[0;1;32m                        ^
[0mIn file included from platform/x11/godot_x11.cpp:37:
In file included from platform/x11/os_x11.h:34:
In file included from platform/x11/context_gl_x11.h:38:
In file included from ./core/os/os.h:34:
In file included from ./core/engine.h:35:
In file included from ./core/os/main_loop.h:34:
In file included from ./core/os/input_event.h:34:
In file included from ./core/math/transform_2d.h:34:
In file included from ./core/math/rect2.h:34:
In file included from ./core/math/vector2.h:34:
[1m./core/math/math_funcs.h:284:77: [0m[0;1;31merror: [0m[1mexpected unqualified-id[0m
  ...float randf() { return (float)rand() / (float)Math::RANDOM_MAX; }
[0;1;32m                                                         ^
[0m[1m/usr/include/stdlib.h:188:20: [0m[0;1;30mnote: [0mexpanded from macro 'RANDOM_MAX'[0m
#define RANDOM_MAX      0x7fffffff      /* (((long)1 << 31) - 1) */
[0;1;32m                        ^
[0m4 errors generated.
scons: *** [platform/x11/godot_x11.x11.tools.64.llvm.o] Error 1
scons: building terminated because of errors.
```
It's not perfect, but it's better than the screenshot form I previously had.

--- a/core/math/math_funcs.h
+++ b/core/math/math_funcs.h
@@ -46,7 +46,8 @@ class Math {
 public:
 	Math() {} // useless to instance
 
-	static const uint64_t RANDOM_MAX = 0xFFFFFFFF;
+	// Not using 'RANDOM_MAX' to avoid conflict with system headers on some OSes (at least NetBSD).
+	static const uint64_t RANDOM_32BIT_MAX = 0xFFFFFFFF;
 
 	static _ALWAYS_INLINE_ double sin(double p_x) { return ::sin(p_x); }
 	static _ALWAYS_INLINE_ float sin(float p_x) { return ::sinf(p_x); }
@@ -283,8 +284,8 @@ public:
 	static void randomize();
 	static uint32_t rand_from_seed(uint64_t *seed);
 	static uint32_t rand();
-	static _ALWAYS_INLINE_ double randd() { return (double)rand() / (double)Math::RANDOM_MAX; }
-	static _ALWAYS_INLINE_ float randf() { return (float)rand() / (float)Math::RANDOM_MAX; }
+	static _ALWAYS_INLINE_ double randd() { return (double)rand() / (double)Math::RANDOM_32BIT_MAX; }
+	static _ALWAYS_INLINE_ float randf() { return (float)rand() / (float)Math::RANDOM_32BIT_MAX; }
 
 	static double random(double from, double to);
 	static float random(float from, float to);

--- a/core/math/random_pcg.h
+++ b/core/math/random_pcg.h
@@ -31,11 +31,11 @@
 #ifndef RANDOM_PCG_H
 #define RANDOM_PCG_H
 
-#include <math.h>
-
 #include "core/math/math_defs.h"
 
 #include "thirdparty/misc/pcg.h"
+
+#include <math.h>
 
 #if defined(__GNUC__)
 #define CLZ32(x) __builtin_clz(x)
@@ -67,7 +67,6 @@ class RandomPCG {
 public:
 	static const uint64_t DEFAULT_SEED = 12047754176567800795U;
 	static const uint64_t DEFAULT_INC = PCG_DEFAULT_INC_64;
-	static const uint64_t RANDOM_MAX = 0xFFFFFFFF;
 
 	RandomPCG(uint64_t p_seed = DEFAULT_SEED, uint64_t p_inc = DEFAULT_INC);
 

--- a/drivers/unix/os_unix.cpp
+++ b/drivers/unix/os_unix.cpp
@@ -48,7 +48,7 @@
 #include <mach/mach_time.h>
 #endif
 
-#if defined(__FreeBSD__) || defined(__OpenBSD__)
+#if defined(__FreeBSD__) || defined(__OpenBSD__) || defined(__NetBSD__)
 #include <sys/param.h>
 #include <sys/sysctl.h>
 #endif
@@ -477,7 +477,7 @@ String OS_Unix::get_executable_path() const {
 		return OS::get_executable_path();
 	}
 	return b;
-#elif defined(__OpenBSD__)
+#elif defined(__OpenBSD__) || defined(__NetBSD__)
 	char resolved_path[MAXPATHLEN];
 
 	realpath(OS::get_executable_path().utf8().get_data(), resolved_path);

--- a/drivers/unix/thread_posix.cpp
+++ b/drivers/unix/thread_posix.cpp
@@ -29,16 +29,16 @@
 /*************************************************************************/
 
 #include "thread_posix.h"
-#include "core/script_language.h"
 
 #if (defined(UNIX_ENABLED) || defined(PTHREAD_ENABLED)) && !defined(NO_THREADS)
+
+#include "core/os/memory.h"
+#include "core/safe_refcount.h"
+#include "core/script_language.h"
 
 #ifdef PTHREAD_BSD_SET_NAME
 #include <pthread_np.h>
 #endif
-
-#include "core/os/memory.h"
-#include "core/safe_refcount.h"
 
 static void _thread_id_key_destr_callback(void *p_value) {
 	memdelete(static_cast<Thread::ID *>(p_value));
@@ -126,6 +126,8 @@ Error ThreadPosix::set_name_func_posix(const String &p_name) {
 #ifdef PTHREAD_BSD_SET_NAME
 	pthread_set_name_np(running_thread, p_name.utf8().get_data());
 	int err = 0; // Open/FreeBSD ignore errors in this function
+#elif defined(PTHREAD_NETBSD_SET_NAME)
+	int err = pthread_setname_np(running_thread, "%s", const_cast<char *>(p_name.utf8().get_data()));
 #else
 	int err = pthread_setname_np(running_thread, p_name.utf8().get_data());
 #endif // PTHREAD_BSD_SET_NAME

--- a/platform/linuxbsd/display_server_x11.cpp
+++ b/platform/linuxbsd/display_server_x11.cpp
@@ -3575,7 +3575,12 @@ DisplayServerX11::DisplayServerX11(const String &p_rendering_driver, WindowMode 
 	xrandr_handle = dlopen("libXrandr.so.2", RTLD_LAZY);
 	if (!xrandr_handle) {
 		err = dlerror();
-		fprintf(stderr, "could not load libXrandr.so.2, Error: %s\n", err);
+		// For some arcane reason, NetBSD now ships libXrandr.so.3 while the rest of the world has libXrandr.so.2...
+		// In case this happens for other X11 platforms in the future, let's give it a try too before failing.
+		xrandr_handle = dlopen("libXrandr.so.3", RTLD_LAZY);
+		if (!xrandr_handle) {
+			fprintf(stderr, "could not load libXrandr.so.2, Error: %s\n", err);
+		}
 	} else {
 		XRRQueryVersion(x11_display, &xrandr_major, &xrandr_minor);
 		if (((xrandr_major << 8) | xrandr_minor) >= 0x0105) {

--- a/platform/linuxbsd/platform_config.h
+++ b/platform/linuxbsd/platform_config.h
@@ -31,7 +31,15 @@
 #ifdef __linux__
 #include <alloca.h>
 #endif
-#if defined(__FreeBSD__) || defined(__OpenBSD__)
-#include <stdlib.h>
+
+#if defined(__FreeBSD__) || defined(__OpenBSD__) || defined(__NetBSD__)
+#include <stdlib.h> // alloca
+// FreeBSD and OpenBSD use pthread_set_name_np, while other platforms,
+// include NetBSD, use pthread_setname_np. NetBSD's version however requires
+// a different format, we handle this directly in thread_posix.
+#ifdef __NetBSD__
+#define PTHREAD_NETBSD_SET_NAME
+#else
 #define PTHREAD_BSD_SET_NAME
+#endif
 #endif

--- a/platform/server/platform_config.h
+++ b/platform/server/platform_config.h
@@ -31,10 +31,19 @@
 #if defined(__linux__) || defined(__APPLE__)
 #include <alloca.h>
 #endif
-#if defined(__FreeBSD__) || defined(__OpenBSD__)
-#include <stdlib.h>
+
+#if defined(__FreeBSD__) || defined(__OpenBSD__) || defined(__NetBSD__)
+#include <stdlib.h> // alloca
+// FreeBSD and OpenBSD use pthread_set_name_np, while other platforms,
+// include NetBSD, use pthread_setname_np. NetBSD's version however requires
+// a different format, we handle this directly in thread_posix.
+#ifdef __NetBSD__
+#define PTHREAD_NETBSD_SET_NAME
+#else
 #define PTHREAD_BSD_SET_NAME
 #endif
+#endif
+
 #ifdef __APPLE__
 #define PTHREAD_RENAME_SELF
 #endif

